File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/Callback.java
package com.roxiemobile.networkingapi.network.rest;

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity;
import com.roxiemobile.networkingapi.network.rest.response.RestApiError;

import org.jetbrains.annotations.NotNull;

public interface Callback<Ti, To> {

    /**
     * TODO
     */
    boolean onShouldExecute(@NotNull Call<Ti> call);

    /**
     * TODO
     */
    void onSuccess(@NotNull Call<Ti> call, @NotNull ResponseEntity<To> entity);

    /**
     * TODO
     */
    void onFailure(@NotNull Call<Ti> call, @NotNull RestApiError error);

    /**
     * TODO
     */
    void onCancel(@NotNull Call<Ti> call);
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/Callback.kt
package com.roxiemobile.networkingapi.network.rest

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity
import com.roxiemobile.networkingapi.network.rest.response.RestApiError

interface Callback<Ti, To> {

// MARK: - Methods

    /**
     * TODO
     */
    fun onShouldExecute(call: Call<Ti>): Boolean {
        return true
    }

    /**
     * TODO
     */
    fun onSucceeded(call: Call<Ti>, responseEntity: ResponseEntity<To>) {
        // Do nothing
    }

    /**
     * TODO
     */
    fun onFailed(call: Call<Ti>, restApiError: RestApiError) {
        // Do nothing
    }

    /**
     * TODO
     */
    fun onCancelled(call: Call<Ti>) {
        // Do nothing
    }
}
